README.md
# Popcorn API

[![Build Status](https://travis-ci.org/popcorn-official/popcorn-api.svg?branch=development)](https://travis-ci.org/popcorn-official/popcorn-api)
[![Dependency Status](https://david-dm.org/popcorn-official/popcorn-api.svg)](https://david-dm.org/popcorn-official/popcorn-api)
[![devDependency Status](https://david-dm.org/popcorn-official/popcorn-api/dev-status.svg)](https://david-dm.org/popcorn-official/popcorn-api#info=devDependencies)
[![document](https://popcorn-official.github.io/popcorn-api/badge.svg)](https://popcorn-official.github.io/popcorn-api/source.html)

Popcorn API is developed to make it easier for anyone to create their own version of [Popcorn Time](http://popcorntime.sh). It contains:

- Metadata about movies (taken from Trakt).
- Metadata about TV shows and individual episodes (taken from Trakt).
- Metadata about anime shows (taken from Hummingbird).
- Multiple quality magnet links for every episode.
- Ability to easily filter content to the user's content.
- Add content manually through the CLI.

# Installation

1. Install MongoDB including `mongoexport` and `mongoimport`.
2. Install NodeJS (at least Node v6.3.0 or greater).
3. Clone the repository with: `git clone https://github.com/popcorn-official/popcorn-api.git`.
4. Install `gulp` globally with `[sudo] npm install -g gulp`.
5. Install dependencies of Popcorn API with `cd popcorn-api` and `npm install`.
6. Build the ES5 code with `gulp build`.

# Manual

Checkout the [manual](https://popcorn-official.github.io/popcorn-api/manual/index.html) to find out more about this project. Or checkout the documentation for the API routes [here](http://docs.popcornofficial.apiary.io/).

# License

MIT License

Copyright (c) 2016 - Popcorn API - Released under the [MIT license](LICENSE.txt).

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

--------------------------------------------------------------------------------

**This project and the distribution of this project is not illegal, nor does it violate _any_ DMCA laws. The use of this project, however, may be illegal in your area. Check your local laws and regulations regarding the use of torrents to watch potentially copyrighted content. The maintainers of this project do not condone the use of this project for anything illegal, in any state, region, country, or planet. _Please use at your own risk_.**

--------------------------------------------------------------------------------

Copyright (c) 2016 - Popcorn API - Released under the [MIT license](LICENSE.txt).


CHANGELOG.md
# 2.2.4 - Keep it working - 17 December 2016

 - **New features:**
   - Testing with [Dredd](https://github.com/apiaryio/dredd) in favor of mocha.
   - Better fetching of images.

 - **Bug fixes:**
   - Fixed ExtraTorrent scraping with [extratorrent-api@1.0.9](https://github.com/ChrisAlderson/extratorrent-api).

# 2.2.3 - Can you see me now? - 19 November 2016

- **New features:**
  - Added `themoviedbclient` as a provider for images.

- **Bug fixes:**
  - `trailer` defaults to `null` instead of `false`.

- **Notes:**
  - Moved `/` to `/status`.
  - Removed JSCS for ESLint.

# 2.2.2 - The key is in the words - 26 October 2016

- **Bug fixes:**
  - Removed special characters when searching with keywords [#44](https://github.com/popcorn-official/popcorn-api/issues/44).

- **Notes:**
  - Movie images are now coming from Fanart.tv or OMDB.
  - Show images are now coming from Fanart.tv or TVDB.

# 2.2.1 - That's for later - 23 August 2016

 - **Bug fixes:**
   - Fixed an issue where the API could return an anime with the type `movie`.
   - Fixed an issue where a season of a date based show could be `0`.
   - Fixed an issue where date based shows where not added correctly.

# 2.2.0 - Extra Torrents? - 11 August 2016

 - **New features:**
   - Added extractors to extract torrents and content data from content providers.
   - Added ExtraTorrent as a content provider with the [extratorrent-api](https://github.com/ChrisAlderson/extratorrent-api) module.
   - DateBased shows are now included in the scrape [#8](https://github.com/popcorn-official/popcorn-api/issues/8).
   - Added Nyaa as a content provider for anime with the [nyaa-api-pt](https://github.com/ChrisAlderson/nyaa-api-pt) module.

 - **Notes:**
   - Added tests for the controllers.
   - Changed `num_episodes` to `num_seasons` for the `Anime` model.

# 2.1.0 - You weeaboos - 2 August 2016

 - **New features:**
   - Anime! For all you weeaboos.
   - Get a random TV show at `/random/show`.
   - Import & export collection.

 - **Bug fixes:**
   - Fixed search not working correctly.
   - Fixed MongoDB sorting limitation [#25](https://github.com/popcorn-official/popcorn-api/issues/25).

 - **Notes:**
   - Moved documentation over to ESDoc.
   - Documentation will now be on GitHub.
   - Added Travis CI.
   - Using ES6 classes instead of factory functions.

# 2.0.0 - Whatcha Babbling about? - 20 June 2016

 - **New features:**
   - Get a random movie at `/random/movie`.
   - Added YTS as a movie provider.
   - Movies now support multiple languages.
   - Moved the entire project to ES6 with [Babel](https://babeljs.io/).
   - The API can now be installed globally.
   - Added a command line interface to the API ([See CLI](https://popcorn-official.github.io/popcorn-api/manual/usage.html#global)).
     - Able to add new show torrents through the CLI.
     - Able to add new movie torrents through the CLI.
   - Added support for the Gulp build system ([See Gulp](https://popcorn-official.github.io/popcorn-api/manual/usage.html#gulp)).
   - Documentation can be generated with the `npm run docs`.
   - Added option to start with the forever module (`npm run forever`).

 - **Bug fixes:**
   - Fixed bug where the cron job was not working.
   - Fixed bug where better movie and show torrents did not update.
   - Fixed bug where `num_seasons` was fluctuating.

 - **Notes:**
   - Changed the projection of movies so it can be used with a [popcorntime provider](https://github.com/ChrisAlderson/butter-provider-movies).
   - (HTTP) Logging is done with the Winston library.
   - Added configuration for JSCS linter.
   - Changed license from GPLv3 to MIT.
   - Moved the library for eztv scraping to its own module over [here](https://github.com/ChrisAlderson/eztv-api-pt).
   - Moved the library for kat scraping to its own module over [here](https://github.com/ChrisAlderson/kat-api-pt).
   - Redid the [README.md](README.md).
   - Redid the kat providers.

# 1.1.0 - Wanna catch a movie? - 20 March 2016

 - **New features:**
   - Movies!
    - Scrape movies from kat.cr the same way shows work.
    - Routes to get a list of movies or a specific movie.
   - Added `/logs/error` to see the error log.
   - Added `imdbMap` in `config.js` for correcting imdb ids.

 - **Bug fixes:**
   - Fixed issue where some season based episodes from EZTV where not added (Including [Last Week Tonight with John Oliver](https://eztv.ag/shows/1025/last-week-tonight-with-john-oliver/)).
   - Partially fixed issue with MongoDB limitations to sorting.
   - Status will now be set to `Idle` after scraping is done.

 - **Notes:**
   - Disabled `/shows/search/`, `/shows/update/`, `/shows/last_updated` routes as they don't seem to be used by Popcorn Time.
   - Made scraping EZTV faster by merging the `getShowDetails` and `getAllEpisodes` functions.
   - Required NodeJS version was changed in 1.0.2 to NodeJS v.5.0.0.
   - Fixed several issues with adding content through the CLI.

# 1.0.2 - Wanna retry? - 14 March 2016

 - **New features:**
   - Resets the log files on each scrape.
   - Added `repo` to the index.

 - **Bug fixes:**
   - Now properly updates metadata.
   - Some fixes to prevent ETIMEDOUT.

 - **Notes:**
   - Removed dependency on Q.
   - Replaced `slug` with `imdb` for getting seasonal metadata from trakt.tv.

# 1.0.1 - What's trending? - 6 March 2016

 - **Bug fixes:**
   - Sort by trending.

# 1.0.0 - Let's kick some ass! - 1 March 2016

 - **Features:**
   - Scraping EZTV.ag just like the old API.
   - Scraping kat.cr with 17 different providers.
   - Able to add more providers for kat.cr scraping.


lib/kat.js
const cheerio = require("cheerio"),
  config = require("../config"),
  util = require("../util");

const BASE_URLS = ["https://kat.cr/usearch/", "https://kickassto.co/usearch/"];
let request = require("request");
request = request.defaults({
  "headers": {
    "Accept-Encoding": "gzip, deflate"
  },
  "gzip": true,
  "timeout": config.webRequestTimeout * 1000
});

const platformMap = {
  "android": 4,
  "blackberry": 7,
  "gamecube": 15,
  "ipad": 18,
  "iphone": 19,
  "ipod": 20,
  "java": 22,
  "linux": 24,
  "mac": 25,
  "nintendo3-ds": 31,
  "nintendo-ds": 33,
  "dvd": 35,
  "other": 65,
  "palm-os": 37,
  "pc": 38,
  "ps2": 43,
  "ps3": 44,
  "ps4": 66,
  "psp": 45,
  "symbian": 52,
  "wii": 56,
  "wiiu": 68,
  "windows-ce": 57,
  "windows-mobile": 58,
  "windows-phone": 59,
  "xbox": 61,
  "xbox-360": 62,
  "xbox-one": 67
};

const languageMap = {
  "en": 2,
  "sq": 42,
  "ar": 7,
  "eu": 44,
  "bn": 46,
  "pt-br": 39,
  "bg": 37,
  "yue": 45,
  "ca": 47,
  "zh": 10,
  "hr": 34,
  "cs": 32,
  "da": 26,
  "nl": 8,
  "tl": 11,
  "fi": 31,
  "fr": 5,
  "de": 4,
  "el": 30,
  "he": 25,
  "hi": 6,
  "hu": 27,
  "it": 3,
  "ja": 15,
  "kn": 49,
  "ko": 16,
  "lt": 43,
  "ml": 21,
  "cmn": 23,
  "ne": 48,
  "no": 19,
  "fa": 33,
  "pl": 9,
  "pt": 17,
  "pa": 35,
  "ro": 18,
  "ru": 12,
  "sr": 28,
  "sl": 36,
  "es": 14,
  "sv": 20,
  "ta": 13,
  "te": 22,
  "th": 24,
  "tr": 29,
  "uk": 40,
  "vi": 38
};

/* Formats the info from a given search page. */
const formatPage = (response, page, date) => {
  const $ = cheerio.load(response);

  const matcher = /\s+[a-zA-Z]+\s\d+[-]\d+\s[a-zA-Z]+\s(\d+)/;
  const totalResults = $("table#mainSearchTable.doublecelltable").find("h2").find("span").text().match(matcher);
  const totalPages = $("div.pages.botmarg5px.floatright").children("a.turnoverButton.siteButton.bigButton").last().text();

  const formatted = {
    response_time: parseInt(date),
    page: parseInt(page),
    totalResults: parseInt(totalResults[1]),
    totalPages: totalPages ? totalPages : 1,
    results: []
  };

  $("table.data").find("tr[id]").each(function() {
    const torrent = {
      title: $(this).find("a.cellMainLink").text(),
      category: $(this).find("span.font11px.lightgrey.block").find("a[href]").last().text(),
      link: $(this).find("a.cellMainLink[href]").attr("href"),
      guid: $(this).find("a.cellMainLink[href]").attr("href"),
      verified: $(this).find("i.ka.ka16.ka-verify.ka-green").length,
      comments: parseInt($(this).find("a.icommentjs.kaButton.smallButton.rightButton").text()),
      magnet: $(this).find("a.icon16[data-nop]").attr("href"),
      torrentLink: $(this).find("a.icon16[data-download]").attr("href"),
      size: parseInt($(this).find("td.center").eq(0).text()),
      files: parseInt($(this).find("td.center").eq(1).text()),
      pubDate: Number(new Date($(this).find("td.center").eq(2).attr("title"))),
      seeds: parseInt($(this).find("td.center").eq(3).text()),
      leechs: parseInt($(this).find("td.center").eq(4).text())
    };
    torrent.peers = torrent.seeds + torrent.leechs;
    formatted.results.push(torrent);
  });

  return formatted;
};

/* Request the data from kat.cr with an endpoint. */
const requestData = (url, endpoint, retry = true) => {
  return new Promise((resolve, reject) => {
    request(url + endpoint, (err, res, body) => {
      if (err && retry) {
        resolve(requestData(BASE_URLS[1], endpoint, false));
      } else if (err) {
        reject(err + " with link: '" + endpoint + "'");
      } else if (!body || res.statusCode >= 400) {
        reject("KAT: Could not load data from: '" + endpoint + "'");
      } else {
        resolve(body);
      }
    });
  });
};

/* Makes an endpoint for the request. */
const makeEndpoint = (query) => {
  let endpoint = "";

  if (!query) {
    util.onError("Field 'query' is required.");
  } else if (typeof(query) === "string") {
    endpoint += query;
  } else if (typeof(query) === "object") {
    if (query.query) endpoint += query.query;
    if (query.category) endpoint += " category:" + query.category;
    if (query.uploader) endpoint += " user:" + query.uploader;
    if (query.min_seeds) endpoint += " seeds:" + query.min_seeds;
    if (query.age) endpoint += " age:" + query.age;
    if (query.min_files) endpoint += " files:" + query.min_files;
    if (query.imdb) endpoint += " imdb:" + query.imdb.replace(/\D/g, "");
    if (query.tvrage) endpoint += " tv:" + query.tvrage;
    if (query.isbn) endpoint += " isbn:" + query.isbn;
    if (query.language) {
      const languageCode = languageMap[query.language] != undefined ? languageMap[query.language] : "";
      endpoint += " lang_id:" + languageCode;
    }
    if (query.adult_filter) endpoint += " is_safe:" + query.adult_filter;
    if (query.verified) endpoint += " verified:" + query.verified;
    if (query.season) endpoint += " season:" + query.season;
    if (query.episode) endpoint += " episode:" + query.episode;
    if (query.platform_id) {
      const platformCode = platformMap[query.platform_id] != undefined ? platformMap[query.platform_id] : "";
      endpoint += " platform_id:" + platformCode;
    }
    if (query.page) endpoint += "/" + query.page;
    if (query.sort_by) endpoint += "/?field=" + query.sort_by;
    if (query.order) endpoint += "&order=" + query.order;
  } else {
    util.onError("No valid query.");
  }

  return endpoint;
};

module.exports = {

  /* Returns the formated data from a search request. */
  search: (query) => {
    const endpoint = makeEndpoint(query);
    const t = Date.now();
    return requestData(BASE_URLS[0], endpoint).then((response) => {
      return formatPage(response, query.page || 1, Date.now() - t);
    });
  }

};
